File: MyMusic.Mobile/src/services/sync/context.ts
import type {SyncContext, ISyncConfig, ISyncState, SyncResult} from './types';
import {decodeToFsPath} from '../pathUtils';

export function createSyncContext(
    config: ISyncConfig,
    state: ISyncState
): SyncContext {
    const deviceId = config.getDeviceId()!;
    const repositoryPath = config.getRepositoryPath();
    const decodedRepoPath = decodeToFsPath(repositoryPath);

    const result: SyncResult = {
        createRemote: 0,
        updateRemote: 0,
        createLocal: 0,
        updateLocal: 0,
        delete: 0,
        link: 0,
        unlink: 0,
        rename: 0,
        skipped: 0,
        conflict: 0,
        updateTimestamp: 0,
        error: 0,
    };

    return {
        deviceId,
        repositoryPath,
        decodedRepoPath,
        options: state.options,
        result,
        uploadedPaths: new Set(),
        pendingDownloadPaths: new Set(),
        conflictedSongIds: new Set(),
    };
}

File: MyMusic.CLI/Services/Sync/Context.cs
namespace MyMusic.CLI.Services.Sync;

using MyMusic.Common.Services.Sync;
using MyMusic.Common.Services.Sync.Types;

public record SyncContext
{
    public long DeviceId { get; init; }
    public required string RepositoryPath { get; init; }
    public long SessionId { get; set; }
    public SyncOptions Options { get; init; } = new();
    public SyncResult Result { get; set; } = new();
    public HashSet<string> UploadedPaths { get; } = [];
    public HashSet<string> PendingDownloadPaths { get; set; } = [];
    public List<PendingActionItem> PendingActions { get; set; } = [];
}
